File: ts/packages/nonces/src/rpc.ts
import type { PublicClient } from "viem";
import { INONCE_ABI, NONCE_PRECOMPILE } from "./constants";

/**
 * Query the Nonce precompile for a specific (address, key) pair.
 *
 * @param client - viem PublicClient instance
 * @param address - Account address to query nonce for
 * @param key - Nonce key (lane index or maxUint256 for expiring)
 * @returns Current nonce value as bigint
 */
export async function getNonceFromPrecompile(
  client: PublicClient,
  address: `0x${string}`,
  key: bigint,
): Promise<bigint> {
  return client.readContract({
    address: NONCE_PRECOMPILE,
    abi: INONCE_ABI,
    functionName: "getNonce",
    args: [address, key],
  });
}

/**
 * Query the protocol nonce (key=0) via eth_getTransactionCount.
 * Utility for callers who need the standard EVM nonce — not used by NoncePool internally.
 *
 * @param client - viem PublicClient instance
 * @param address - Account address
 * @returns Pending nonce as bigint
 */
export async function getProtocolNonce(
  client: PublicClient,
  address: `0x${string}`,
): Promise<bigint> {
  const count = await client.getTransactionCount({
    address,
    blockTag: "pending",
  });
  return BigInt(count);
}

File: rs/nonces/src/rpc.rs
use alloy::primitives::{Address, U256};
use alloy::providers::Provider;
use alloy::sol;

use crate::constants::NONCE_PRECOMPILE;
use crate::error::NonceError;

sol! {
    #[sol(rpc)]
    interface INonce {
        function getNonce(address owner, uint256 key) external view returns (uint64);
    }
}

/// Query the nonce precompile for a specific (address, key) pair.
pub async fn get_nonce_from_precompile<P: Provider>(
    provider: &P,
    address: Address,
    key: U256,
) -> Result<u64, NonceError> {
    let contract = INonce::new(NONCE_PRECOMPILE, provider);
    let nonce = contract.getNonce(address, key).call().await?;
    Ok(nonce)
}

/// Utility for callers who need the standard EVM nonce -- not used by NoncePool internally.
///
/// Queries the protocol nonce (transaction count) for an address using the `pending` block tag.
pub async fn get_protocol_nonce<P: Provider>(
    provider: &P,
    address: Address,
) -> Result<u64, NonceError> {
    let count = provider
        .get_transaction_count(address)
        .pending()
        .await
        .map_err(|e| NonceError::Rpc(alloy::contract::Error::TransportError(e)))?;
    Ok(count)
}
